File: backend/internal/devmock/user_pokemon_repo.go
package devmock

import (
	"context"
	"fmt"
	"sort"
	"sync"
	"time"

	"github.com/kenyamamoto/pokelingual/backend/internal/model"
)

// UserPokemonRepo implements domain.UserPokemonRepository with in-memory storage.
type UserPokemonRepo struct {
	mu   sync.Mutex
	data map[string]map[int]*model.UserPokemon // uid -> pokemonID -> record
}

// NewUserPokemonRepo creates a new in-memory UserPokemonRepo.
func NewUserPokemonRepo() *UserPokemonRepo {
	return &UserPokemonRepo{
		data: make(map[string]map[int]*model.UserPokemon),
	}
}

// UpsertEncounter creates or updates a user's encounter record for a Pokemon.
func (r *UserPokemonRepo) UpsertEncounter(ctx context.Context, uid string, pokemonID int, score float64, captured bool) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.data[uid] == nil {
		r.data[uid] = make(map[int]*model.UserPokemon)
	}

	now := time.Now()
	existing, ok := r.data[uid][pokemonID]
	if !ok {
		status := "seen"
		if captured {
			status = "captured"
		}
		up := &model.UserPokemon{
			PokemonID:         pokemonID,
			Status:            status,
			TotalCaptures:     boolToInt(captured),
			TotalEncounters:   1,
			LastEncounteredAt: now,
			BestScore:         score,
		}
		if captured {
			up.LastCapturedAt = now
		}
		r.data[uid][pokemonID] = up
		return nil
	}

	existing.TotalEncounters++
	existing.LastEncounteredAt = now
	if captured {
		existing.TotalCaptures++
		existing.Status = "captured"
		existing.LastCapturedAt = now
	}
	if score > existing.BestScore {
		existing.BestScore = score
	}
	return nil
}

// GetCollection returns all discovered Pokemon for the given user, ordered by Pokemon ID.
func (r *UserPokemonRepo) GetCollection(ctx context.Context, uid string) ([]model.UserPokemon, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	result := make([]model.UserPokemon, 0)
	for _, up := range r.data[uid] {
		result = append(result, *up)
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].PokemonID < result[j].PokemonID
	})
	return result, nil
}

// GetPokemon returns a specific Pokemon record for the given user.
func (r *UserPokemonRepo) GetPokemon(ctx context.Context, uid string, pokemonID int) (*model.UserPokemon, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.data[uid] == nil {
		return nil, fmt.Errorf("pokemon not found")
	}
	up, ok := r.data[uid][pokemonID]
	if !ok {
		return nil, fmt.Errorf("pokemon not found")
	}
	return up, nil
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

File: backend/src/devmock/user-pokemon-repo.ts
import type { UserPokemonRepository } from "../domain/interfaces.js";
import type { UserPokemon } from "../types/index.js";

/** ローカル開発用のインメモリ UserPokemonRepository 実装。 */
export class MockUserPokemonRepo implements UserPokemonRepository {
  private data = new Map<string, Map<number, UserPokemon>>();

  async upsertEncounter(uid: string, pokemonID: number, score: number, captured: boolean): Promise<void> {
    if (!this.data.has(uid)) this.data.set(uid, new Map());
    const userMap = this.data.get(uid)!;

    const now = new Date();
    const existing = userMap.get(pokemonID);

    if (!existing) {
      userMap.set(pokemonID, {
        pokemon_id: pokemonID,
        status: captured ? "captured" : "seen",
        total_captures: captured ? 1 : 0,
        total_encounters: 1,
        last_captured_at: captured ? now : null,
        last_encountered_at: now,
        best_score: score,
      });
      return;
    }

    existing.total_encounters++;
    existing.last_encountered_at = now;
    if (captured) {
      existing.total_captures++;
      existing.status = "captured";
      existing.last_captured_at = now;
    }
    if (score > existing.best_score) {
      existing.best_score = score;
    }
  }

  async getCollection(uid: string): Promise<UserPokemon[]> {
    const userMap = this.data.get(uid);
    if (!userMap) return [];
    return [...userMap.values()].sort((a, b) => a.pokemon_id - b.pokemon_id);
  }

  async getPokemon(uid: string, pokemonID: number): Promise<UserPokemon> {
    const userMap = this.data.get(uid);
    const pokemon = userMap?.get(pokemonID);
    if (!pokemon) throw new Error("pokemon not found");
    return pokemon;
  }
}
